apps/builder/app/builder/features/ai/copywriter.tsx

```diff
@@ -0,0 +1,191 @@
+import { z } from "zod";
+import store from "immerhin";
+import untruncateJson from "untruncate-json";
+import { copywriter, requestStream } from "@webstudio-is/ai";
+import { restAiCopy } from "~/shared/router-utils";
+import {
+  Box,
+  Button,
+  InputField,
+  Label,
+  Text,
+} from "@webstudio-is/design-system";
+import {
+  instancesStore,
+  projectStore,
+  selectedInstanceStore,
+} from "~/shared/nano-states";
+import { useStore } from "@nanostores/react";
+import { collectTextInstances } from "node_modules/@webstudio-is/ai/src/chains/copywriter";
+import { useRef, useState } from "react";
+import { computed } from "nanostores";
+
+const patchTextInstance = (textInstance: copywriter.TextInstance) => {
+  store.createTransaction([instancesStore], (instances) => {
+    const currentInstance = instances.get(textInstance.instanceId);
+
+    if (
+      currentInstance === undefined ||
+      currentInstance.children.length === 0
+    ) {
+      return;
+    }
+
+    // Instances can have a number of text child nodes without interleaving components.
+    // When this is the case we treat the child nodes as a single text node,
+    // otherwise the AI would generate children.length chunks of separate text.
+    // We can identify this case of "joint" text instances when the index is -1.
+    const replaceAll = textInstance.index === -1;
+    if (replaceAll) {
+      if (currentInstance.children.every((child) => child.type === "text")) {
+        currentInstance.children = [{ type: "text", value: textInstance.text }];
+      }
+      return;
+    }
+
+    if (currentInstance.children[textInstance.index].type === "text") {
+      currentInstance.children[textInstance.index].value = textInstance.text;
+    }
+  });
+};
+
+const onChunk = (completion: string) => {
+  try {
+    const jsonResponse = z
+      .array(copywriter.TextInstanceSchema)
+      .parse(JSON.parse(untruncateJson(completion)));
+
+    const currenTextInstance = jsonResponse.pop();
+
+    if (currenTextInstance === undefined) {
+      return;
+    }
+
+    patchTextInstance(currenTextInstance);
+  } catch (error) {
+    /**/
+  }
+};
+
+export const Copywriter = () => {
+  const [isLoading, setIsLoading] = useState(false);
+  const abort = useRef<AbortController | null>(null);
+
+  const project = useStore(projectStore);
+
+  const textInstances = computed(
+    [instancesStore, selectedInstanceStore],
+    (instances, selectedInstance) => {
+      if (selectedInstance) {
+        return collectTextInstances({
+          instances: instancesStore.get(),
+          rootInstanceId: selectedInstance.id,
+        }).map((instance) => {
+          if (instance.text === "Text you can edit") {
```

Comment:
Very flaky assumption

Reply:
how do we detect these? maybe we should have used a placeholder property with an empty text property? or something

Reply:
Otherwise we need to extract text to some const or meta and compare to see if its a default

Reply:
I am sure @gsppe just doesn't have a solution for doing this well and we could help out

Reply:
could be instance.text || instance.placeholder for rendering, then this check would be basically `if (instance.text)`

Reply:
we are discussing it with @TrySound, ignore it for now

Reply:
We decided to remove placeholders from data so you can safely remove this conversion too.
https://github.com/webstudio-is/webstudio/pull/2399